### PR TITLE
App Config Spring - Arch Review Comments

### PR DESCRIPTION
**Pull request by mrm9084**
# Description

Updates from the Arch Review + bug fixes.

# All SDK Contribution checklist:
- [ ] **The pull request does not introduce [breaking changes]**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [ ] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md).**

## [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#developer-guide)
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#building-and-unit-testing)
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/spring/spring-cloud-azure-feature-management-web/src/main/java/com/azure/spring/cloud/feature/manager/FeatureManagerSnapshot.java [hunk 1]
```diff
@@ -44,4 +44,24 @@ public Mono<Boolean> isEnabledAsync(String feature) {
 
         return featureManager.isEnabledAsync(feature).doOnSuccess((enabled) -> requestMap.put(feature, enabled));
     }
+
+    /**
+     * Checks to see if the feature is enabled. If enabled it check each filter, once a single filter returns true it
```
Comment (chenrujun) [inline sdk/spring/spring-cloud-azure-feature-management-web/src/main/java/com/azure/spring/cloud/feature/manager/FeatureManagerSnapshot.java:49]: > If enabled it check each filter, once a single filter returns true ...

Seems this content is related to current method. How about delete it?

File: sdk/spring/spring-cloud-azure-feature-management-web/src/main/java/com/azure/spring/cloud/feature/manager/FeatureManagerSnapshot.java [hunk 2]
```diff
@@ -44,4 +44,24 @@ public Mono<Boolean> isEnabledAsync(String feature) {
...
+     * <p>
+     * If isEnabled has already been called on this feature in this request, it will return the same value as it did
+     * before.
+     *
+     * @param feature Feature being checked.
+     * @return state of the feature
+     */
+    public Boolean isEnabled(String feature) {
+        if (requestMap.get(feature) != null) {
+            return requestMap.get(feature);
```
Comment (chenrujun) [inline sdk/spring/spring-cloud-azure-feature-management-web/src/main/java/com/azure/spring/cloud/feature/manager/FeatureManagerSnapshot.java:61]: Avoid call `requestMap.get(feature)` twice.

File: sdk/spring/spring-cloud-azure-feature-management/src/main/java/com/azure/spring/cloud/feature/manager/FeatureManager.java [hunk 3]
```diff
@@ -57,9 +58,22 @@ public class FeatureManager {
      * @return state of the feature
      * @throws FilterNotFoundException file not found
      */
-    public Mono<Boolean> isEnabledAsync(String feature) throws FilterNotFoundException {
+    public Mono<Boolean> isEnabledAsync(String feature) {
         return Mono.just(checkFeatures(feature));
     }
+    
+    /**
+     * Checks to see if the feature is enabled. If enabled it check each filter, once a single filter returns true it
```
Comment (chenrujun) [inline sdk/spring/spring-cloud-azure-feature-management/src/main/java/com/azure/spring/cloud/feature/manager/FeatureManager.java:66]: Same here.

File: sdk/spring/spring-cloud-azure-feature-management/src/main/java/com/azure/spring/cloud/feature/manager/FeatureManager.java [hunk 4]
```diff
@@ -57,9 +58,22 @@ public class FeatureManager {
...
+     *
+     * @param feature Feature being checked.
+     * @return state of the feature
+     * @throws FilterNotFoundException file not found
+     */
+    public Boolean isEnabled(String feature) throws FilterNotFoundException {
+        return checkFeatures(feature);
+    }
 
     private boolean checkFeatures(String feature) throws FilterNotFoundException {
```
Comment (chenrujun) [inline sdk/spring/spring-cloud-azure-feature-management/src/main/java/com/azure/spring/cloud/feature/manager/FeatureManager.java:78]: How about rename this method to make it easier to understand. For example: `featureIsOnAndPassedFilter`

File: sdk/spring/spring-cloud-azure-feature-management/src/main/java/com/azure/spring/cloud/feature/manager/FeatureManager.java [hunk 5]
```diff
@@ -57,9 +58,22 @@ public class FeatureManager {
...
+    /**
+     * Checks to see if the feature is enabled. If enabled it check each filter, once a single filter returns true it
+     * returns true. If no filter returns true, it returns false. If there are no filters, it returns true. If feature
+     * isn't found it returns false.
+     *
+     * @param feature Feature being checked.
+     * @return state of the feature
+     * @throws FilterNotFoundException file not found
+     */
+    public Boolean isEnabled(String feature) throws FilterNotFoundException {
```
Comment (chenrujun) [inline sdk/spring/spring-cloud-azure-feature-management/src/main/java/com/azure/spring/cloud/feature/manager/FeatureManager.java:74]: When will this method throw `FilterNotFoundException`? Is it better to return `true/false` when `FilterNotFoundException` been throwed?

File: sdk/spring/spring-cloud-azure-feature-management/src/main/java/com/azure/spring/cloud/feature/manager/FeatureManager.java [hunk 6]
```diff
@@ -81,7 +95,7 @@ private boolean checkFeatures(String feature) throws FilterNotFoundException {
 
         return featureItem.getEnabledFor().values().stream().filter(Objects::nonNull)
             .filter(featureFilter -> featureFilter.getName() != null)
-            .map(featureFilter -> isFeatureOn(featureFilter, feature)).findAny().orElse(false);
+            .map(featureFilter -> isFeatureOn(featureFilter, feature)).anyMatch(value -> value);
```
Comment (chenrujun) [inline sdk/spring/spring-cloud-azure-feature-management/src/main/java/com/azure/spring/cloud/feature/manager/FeatureManager.java:98]: Change
> .map(featureFilter -> isFeatureOn(featureFilter, feature)).anyMatch(value -> value);

to 
> .anyMatch(featureFilter -> isFeatureOn(featureFilter, feature));

File: sdk/spring/spring-cloud-azure-feature-management/src/main/java/com/azure/spring/cloud/feature/manager/filters/TargetingFilter.java [hunk 7]
```diff
@@ -99,24 +99,27 @@ public boolean evaluate(FeatureFilterEvaluationContext context) {
             throw new IllegalArgumentException("Targeting Context not configured.");
         }
 
-        TargetingContext targetingContext = contextAccessor.getContextAsync().block();
+        TargetingFilterContext targetingContext = new TargetingFilterContext();
 
-        if (targetingContext == null) {
+        contextAccessor.getContextAsync(targetingContext);
+
+        if (!StringUtils.hasText(targetingContext.getUserId()) && !(targetingContext.getGroups() != null
```
Comment (chenrujun) [inline sdk/spring/spring-cloud-azure-feature-management/src/main/java/com/azure/spring/cloud/feature/manager/filters/TargetingFilter.java:106]: Convert this to a private method to make it easier to read.

File: sdk/spring/spring-cloud-azure-feature-management/src/main/java/com/azure/spring/cloud/feature/manager/targeting/TargetingContextAccessor.java [hunk 8]
```diff
@@ -2,17 +2,15 @@
...
-import reactor.core.publisher.Mono;
-
 /**
  * Interface for providing a Targeting Context to the `Microsoft.TargetingFilter`.
  */
-public interface ITargetingContextAccessor {
+public interface TargetingContextAccessor {
 
     /**
      * Returns the Targeting Context for Feature Targeting evaluation.
```
Comment (chenrujun) [inline sdk/spring/spring-cloud-azure-feature-management/src/main/java/com/azure/spring/cloud/feature/manager/targeting/TargetingContextAccessor.java:11]: No `returns` anymore.

File: sdk/spring/spring-cloud-azure-starter-appconfiguration-config/pom.xml [hunk 9]
```diff
@@ -14,6 +14,7 @@
   <artifactId>spring-cloud-azure-starter-appconfiguration-config</artifactId>
   <version>4.0.0-beta.2</version> <!-- {x-version-update;com.azure.spring:spring-cloud-azure-starter-appconfiguration-config;current} -->
   <name>Azure Spring Cloud Starter App Configuration Config</name>
```
Comment (chenrujun) [inline sdk/spring/spring-cloud-azure-starter-appconfiguration-config/pom.xml:16]: `Azure Spring Cloud` -> `Spring Cloud Azure`. Please check all other places.

**Review by chenrujun — COMMENTED** [review]
Please fix the pipeline failure.

File: sdk/spring/spring-cloud-azure-feature-management-web/src/main/java/com/azure/spring/cloud/feature/manager/FeatureManagerSnapshot.java (hunk 1 above)
Comment (mrm9084) [reply sdk/spring/spring-cloud-azure-feature-management-web/src/main/java/com/azure/spring/cloud/feature/manager/FeatureManagerSnapshot.java:49]: I am not sure what you mean by this.

File: sdk/spring/spring-cloud-azure-feature-management/src/main/java/com/azure/spring/cloud/feature/manager/FeatureManager.java (hunk 4 above)
Comment (mrm9084) [reply sdk/spring/spring-cloud-azure-feature-management/src/main/java/com/azure/spring/cloud/feature/manager/FeatureManager.java:78]: I dislike giving it a longer name as doesn't tell the whole story. Would rather make it a JavaDoc. Changing it to `checkFeature` though as it only checks 1 at a time.

File: sdk/spring/spring-cloud-azure-feature-management/src/main/java/com/azure/spring/cloud/feature/manager/FeatureManager.java (hunk 5 above)
Comment (mrm9084) [reply sdk/spring/spring-cloud-azure-feature-management/src/main/java/com/azure/spring/cloud/feature/manager/FeatureManager.java:74]: All filters are defined as Spring Beans, if the filter isn't found then we throw an error as they should be defined. This logic is the same across our Spring and DotNet libraries.

File: sdk/spring/spring-cloud-azure-appconfiguration-config/src/main/java/com/azure/spring/cloud/appconfiguration/config/implementation/AppConfigurationRefreshUtil.java [hunk 10]
```diff
@@ -239,7 +239,7 @@ private static void refreshWithTimeFeatureFlags(AppConfigurationReplicaClient cl
             StateHolder.getCurrentState().updateStateRefresh(state, refreshInterval);
         }
     }
-
+
```
Comment (chenrujun) [inline sdk/spring/spring-cloud-azure-appconfiguration-config/src/main/java/com/azure/spring/cloud/appconfiguration/config/implementation/AppConfigurationRefreshUtil.java:242]: nit: Avoid add useless spaces.

File: sdk/spring/spring-cloud-azure-feature-management/src/main/java/com/azure/spring/cloud/feature/manager/FeatureManager.java (hunk 5 above)
Comment (chenrujun) [reply sdk/spring/spring-cloud-azure-feature-management/src/main/java/com/azure/spring/cloud/feature/manager/FeatureManager.java:74]: Got it. Thanks.

File: sdk/spring/spring-cloud-azure-feature-management/src/main/java/com/azure/spring/cloud/feature/manager/targeting/TargetingContextAccessor.java [hunk 11]
```diff
@@ -2,17 +2,15 @@
...
+public interface TargetingContextAccessor {
 
     /**
-     * Returns the Targeting Context for Feature Targeting evaluation.
-     * @return Mono{@literal <}TargetContext{@literal >}
+     * Configures the Targeting Context for Feature Targeting evaluation.
+     * @param context Targeting Context for selecting feature
      */
-    Mono<TargetingContext> getContextAsync();
+    void getContextAsync(TargetingContext context);
```
Comment (chenrujun) [inline sdk/spring/spring-cloud-azure-feature-management/src/main/java/com/azure/spring/cloud/feature/manager/targeting/TargetingContextAccessor.java:14]: How about renaming to `configureTergetingContextAsync`?

File: eng/code-quality-reports/src/main/resources/revapi/revapi.json [hunk 12]
```diff
@@ -427,6 +427,12 @@
           "code": "java\\.annotation\\.removed",
           "old": ".*? com\\.azure\\.search\\.documents(\\.indexes)?\\.models.*",
           "justification": "Removing Jackson annotations from Search in transition to stream-style."
+        },
+        {
+          "regex": true,
+          "code": "java\\.class\\.externalClassExposedInAPI",
+          "new": "(interface|enum) com\\.azure\\.spring\\.cloud\\.appconfiguration\\.config.*",
+          "justification": "Public class exposed in API"
```
Comment (chenrujun) [inline eng/code-quality-reports/src/main/resources/revapi/revapi.json:435]: Which one is exposed?

**Review by chenrujun — COMMENTED** [review]
Please fix the pipeline failure.

File: sdk/spring/spring-cloud-azure-feature-management-web/src/main/java/com/azure/spring/cloud/feature/manager/FeatureManagerSnapshot.java (hunk 1 above)
Comment (chenrujun) [reply sdk/spring/spring-cloud-azure-feature-management-web/src/main/java/com/azure/spring/cloud/feature/manager/FeatureManagerSnapshot.java]: np, ignore my previous comment.

File: sdk/spring/spring-cloud-azure-feature-management/src/main/java/com/azure/spring/cloud/feature/manager/targeting/TargetingContextAccessor.java (hunk 11 above)
Comment (mrm9084) [reply sdk/spring/spring-cloud-azure-feature-management/src/main/java/com/azure/spring/cloud/feature/manager/targeting/TargetingContextAccessor.java:14]: I was trying to limit the number of breaking change the method has been out like that for at least 2ish years.

File: eng/code-quality-reports/src/main/resources/revapi/revapi.json (hunk 12 above)
Comment (mrm9084) [reply eng/code-quality-reports/src/main/resources/revapi/revapi.json:435]: `AppConfigurationRefresh` and `AppCOnfigurationStoreHealth`. Both are used by `spring-cloud-azure-actuator-autoconfigure`.

File: sdk/spring/spring-cloud-azure-feature-management/src/main/java/com/azure/spring/cloud/feature/manager/targeting/TargetingContextAccessor.java (hunk 11 above)
Comment (chenrujun) [reply sdk/spring/spring-cloud-azure-feature-management/src/main/java/com/azure/spring/cloud/feature/manager/targeting/TargetingContextAccessor.java:14]: OK.

Comment (chenrujun) [reply sdk/spring/spring-cloud-azure-feature-management/src/main/java/com/azure/spring/cloud/feature/manager/targeting/TargetingContextAccessor.java:14]: nit:
1. We already updated the interface name. It already a breaking change. Change method is not a big change comparing to change interface name.
2. `getContextAsync` will make customer confused that it returns `void` but havs `get` in the method name.

Of course, you can decide by yourself. I just give some suggestion.

File: eng/code-quality-reports/src/main/resources/revapi/revapi.json (hunk 12 above)
Comment (mrm9084) [reply eng/code-quality-reports/src/main/resources/revapi/revapi.json:435]: This happens on any library that uses them in my experience. I have had to make updates for the SDK packages too.

Comment (mrm9084) [reply eng/code-quality-reports/src/main/resources/revapi/revapi.json]: I have updated to the new way to do this. The package name change broke it.

**Review by chenrujun — APPROVED** [review]
LGTM.

**Comment by chenrujun** [comment]
@mrm9084 The failure still exists. It only fails on `ubuntu2004_117_netty`.

**Pull request merged**